Add note about how translations work (#31853)
dimagi/commcare-hq: docs/translations.rst

`Django Docs translations page <https://docs.djangoproject.com/en/dev/topics/i18n/translation/>`_
or `this helpful blog post <http://blog.bessas.me/post/65775299341/using-gettext-in-django>`_.

How translations work
---------------------
There are three major components to the translations system - identifying
strings that need to be translated, having human translators write translations
for those strings, and then actually inserting those translations when
appropriate.

> 1. ``./manage.py makemessages`` goes over the source code and pulls out strings

Reviewer: ```suggestion
1. ``./manage.py makemessages`` scans the source code and pulls out strings
```

   that have been tagged for translation. It does not actually execute code, so

>    you really can't do anything fancy in these strings.

Reviewer 2: Maybe this?

```suggestion
   you must not format the string _inside_ the translation lookup function (i.e.
   never use f-strings or call `.format()` inside `gettext(...)`)
```

Reviewer 2: Explicitly "prohibiting" f-strings here would clarify the entire doc, and I'd walk back my _"could be either that or an f-string"_ statement.

Reviewer 2: I don't like my suggestion anymore after seeing your comment above :joy:

>    you really can't do anything fancy in these strings.

Reviewer 2: It might be nice to specify what practices are considered "fancy". An example or two could be helpful, if there's a concise way to describe this, that would be even better.

Author: good point, I'm not really sure.  I think at its core you can really only put strings in here.  Mabe I should rephrase this like
```suggestion
   you can only use simple string literals here.
```
That is, formatting a string won't work, nor will referencing a variable that holds a string.  For example:
```python
MYSTR = "foo"
translated = gettext_noop(MYSTR)
```
This won't actually add any strings to the translation file, because to do so would require evaluating the code.

Reviewer 2: Yeah, I actually totally missed this point in my mind, I was already limiting the scenario where a developer puts an f-string or call "".format() inside the gettext call.

2. Transifex is a third party site we use for translating these strings. Human
   translators look at the strings generated by makemessages and type up
   appropriate translations. They have no context beyond what's in the string
   itself. Be nice to them and give template variables names, so instead of "{}

>    version" they see "{build_date} version", which will be easier to understand.

Reviewer 2: The first example (`instead of "{}...`) is obviously `str.format()` syntax, but the second (`see "{build_date}...`) could be either that or an f-string. If the difference matters, it would be nice to clarify that here.

Reviewer 2: On review review, maybe you were implying `"{build_date} version"` is _not_ an f-string (because you didn't put on a leading `f`)?  I read this as the quotes referring to exactly what the translator sees as the "body of the thing to be translated".

Author: I meant for both to be `str.format`, but that's secondary to the point.  What I wanted to convey was that translators will _only_ see the string template, not the surrounding code or the page.  `"{x} version"` is similarly inscrutable.  Any suggestions for how to phrase that without getting too long-winded here?

Reviewer 2: Hmm, I think if this doc outlines (somewhere earlier) that f-strings must _not_ be used ever (because they will be evaluated prior to translation lookup), then it's fine to leave this as implying `str.format()`.

Author: I updated that text, lemme know what you think

3. Finally, strings are actually translated when the code is executed on the
   production servers. ``gettext`` is a simple function call that takes the
   string provided and looks for it in the translations for the active language.
   If present, it returns the corresponding translation. You can think of it
   like this

   .. code-block:: python

      TRANSLATIONS = {
          "en": {"hello": None},
          "es": {"hello": "hola"},
          "fra": {"hello": "bonjour"},
      }

      def dumb_gettext(str):
          return TRANSLATIONS[get_language()][str] or str

> 

Reviewer: What would you think of adding something like this? I comment on f strings in translations pretty often in PRs.

---

This means that the exact string passed to `gettext` must exist in the translation files. Therefore, f strings cannot be used with `gettext`. Code like this

```
gettext(f"Today is {day_of_week}")
```

will result in a translation file entry like this

```
#: corehq/apps/accounting/filters.py corehq/apps/accounting/forms.py
msgid "Today is {day_of_week}"
msgstr "Hoy es {day_of_week}"
```

but when code executes, the string passed to gettext will be something like "Today is Friday" which will not be found in translations.

Because of this, instead of using f strings with gettext, use `format` on the output of `gettext`: `gettext("Today is {}").format(day_of_week)`

Reviewer: ...I now see this PR originated in a discussion of f strings.

Author: Just added a table with some examples
